Distributor write path: put httpgrpc.Errorf() calls at the topmost level

duricanikolic opened the pull request:
#### What this PR does
This PR introduces the following error types, reppresenting the problems on the distributor’s write path that need some special handling: 
-  `ReplicasNotMatch`, 
- `TooManyClusters`, 
- `ValidationError`, 
- `IngestionRateLimited`
- `RequestRateLimited`.
They all belong to the newly introduced package `distributorerror`.

Moreover, this PR refactors the distributor's write path in such a way that all the calls to `httpgrpc.Errorf()` have been moved to the topmost possible level,i.e., to:
- `push.handler()` and
- `distributor.Push()`.
The semantics of the distributor's write path has not been changed, i.e., the newly introduced errors are mapped to the same HTTP status codes their corresponding problems used to be assigned before this PR.

This refactoring is just a preliminary change needed for a bigger refactoring and improving of intester's and distributor's write path error handling.

#### Which issue(s) this PR fixes or relates to

Related to https://github.com/grafana/mimir/issues/1900, https://github.com/grafana/mimir/issues/6008

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

colega commented on pkg/distributor/distributor.go:
```diff
@@ -43,6 +43,7 @@ import (
 	ingester_client "github.com/grafana/mimir/pkg/ingester/client"
 	"github.com/grafana/mimir/pkg/mimirpb"
 	"github.com/grafana/mimir/pkg/util"
+	distributor_error "github.com/grafana/mimir/pkg/util/error"
```
Several nits. They don't necessarily all apply to the import line, but it's a good place to discuss it IMO:

- Packages shouldn't steal good variable names from the code, same applies to stealing a name as the well known interface `error`.
- If you know that all usages of a package will have to use an alias, it's better to just rename the package to that name, so all usages will be consistent and easier to identify (instead of each usage having a different alias for it)
- Package names (and aliases) should be all lowercase without underscores (`distributorerror`)
- Why distributor errors are under the `util` package? They should be under `distributor`.

So, suggestion:
- Rename `package error` to `package distributorerror` (to comply with golang style guide)
- Move it under `pkg/distributor/distributorerror` (because they belong to distributor, but we don't want to create cyclic dependencies by putting them in `pkg/distributor`)
- Rename `type DistributorError` to `type Error`, so the full name would be `distributorerror.Error` (again, following the golang style guide, to avoid stuttering, as the full name of an entity includes the package name)

duricanikolic replied:
I have implemented these suggestions. Thank you.

pstibrany replied:
I agree with Oleg that these errors should be under distributor package. I would go as far as move them directly into `distributor` package.

To fix the cyclic dependency, I think we should remove dependency from distributor to push package. From what I can see, we only use `push.Func` type. We could easily copy that type into `distributor` package.

Dependency from `push` to `distributor` imho makes sense, since push needs to call distributor or deal with errors. Dependency from `distributor` to `push` should be removed.

duricanikolic replied:
>To fix the cyclic dependency, I think we should remove dependency from distributor to push package. From what I can see, we only use push.Func type. We could easily copy that type into distributor package.

We use `push.Request` too. It is used in `ingester.go`, `ingester_activity.go`, `api.go`. We would need to move that one too if we decide to get rid of the `push` package in `distributor`.
What would be a good place to move `push.Request` and `push.Func`? Somewhere in `util`?

pstibrany replied:
> We use `push.Request` too. It is used in `ingester.go`, `ingester_activity.go`, `api.go`. We would need to move that one too if we decide to get rid of the `push` package in `distributor`. What would be a good place to move `push.Request` and `push.Func`? Somewhere in `util`?

For `push.Request` we could define interface with 3 public methods.

> Somewhere in `util`?

`util/push` seems like a good place for sharing. (ie. where it lives now). Perhaps we could instead move the handler out of `util/push`.

In fact moving push handler under `pkg/distributor` would not be a terrible idea either, I think. The only reason why push handler was in different package is that it was reused between Cortex and Loki. But we don't need to care about that anymore.

duricanikolic replied:
@pstibrany Let's do this additional refactoring in a follow-up PR. Is it OK for you?

pstibrany replied:
I'd suggest trying to do it in preliminary PR, to make changes in this PR cleaner.

colega replied:
I'm not against, but I also don't see the need of moving the errors directly to the `distributor` package right now. If that causes minimal friction, we can keep the `distributorerror` package.

duricanikolic replied:
@pstibrany @colega I would really prefer to do it later. First of all because it might require a big refactoring of the current PR.

duricanikolic replied:
>In fact moving push handler under pkg/distributor would not be a terrible idea either, I think. The only reason why push handler was in different package is that it was reused between Cortex and Loki. But we don't need to care about that anymore.

And what about the otlp handler? Should it also be moved under `pkg/distributor` package?

colega replied:
I think we can refactor moving the push handler later. There's no need to do that now. Especially since ingester also uses the push handler.

pstibrany replied:
That's interesting point. Ingester's version of push handler will never need to deal with Distributor errors ;-)

I wonder if we *need* to have ingesters using push handler in the first place. 🤔

pstibrany replied:
> And what about the otlp handler? Should it also be moved under pkg/distributor package?

I think so. But I can see an argument for `pkg/distributor/handlers` or `pkg/distributor/push` and `pkg/distributor/otlp`.

Perhaps `util/push` could have generic version of push handler, and we can inject error-mapping functions (defined in `distributor` and `ingester` packages) to it? Then distributor and ingester errors don't even need to be public. 🤔

duricanikolic replied:
@pstibrany There is this usage for "testing and debugging": https://github.com/grafana/mimir/blob/main/pkg/api/api.go#L272

colega commented on pkg/distributor/distributor.go:
```diff
@@ -696,12 +694,12 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
 		req, err := pushReq.WriteRequest()
 		if err != nil {
-			return err
+			return distributor_error.NewDistributorPushError(err)
```
What is the point of `NewDistributorPushError` wrapping?

duricanikolic replied:
My idea was to ensure that **all** the errors returned by distributors write path are of type `DistributorError` (that will be renamed in `distributorerror.Error`).

duricanikolic replied:
The only place on distributor's write path where a cast to a `DistributorError` is not done is a place where the errors from ingesters are returned. This will be done in a follow-up PR when ingester errors will be re-organized analogously to this PR.

pstibrany replied:
But this is not distributor error. It's error in parsing the request.

duricanikolic replied:
There are only 2 errors like that:
- `pushReq.WriteRequest()` and
- `tenant.TenantID()`
Do we want to return them as they are? Or do we want to create another error type (e.g., InternalDistributorPushError`) in `distributorerror` package, to be used for wrapping these 2 errors?

colega replied:
As far as I see these errors are treated as InternalServerError, so what's the different between not wrapping them? 

I'd leave the internal errors without wrapping as there's nothing you can do about them and it doesn't add more value. 

As a firstcomer to the code, I expected to see a special behavior associated to this wrapping, but there's none.

pstibrany replied:
> I'd leave the internal errors without wrapping as there's nothing you can do about them and it doesn't add more value.

👍

duricanikolic replied:
I will return the 2 errors as they are, without wrapping them.

colega commented on pkg/distributor/distributor.go:
```diff
@@ -1093,11 +1093,43 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
+		return err
+	}
+	var distributorError distributor_error.DistributorError
+	if errors.As(err, &distributorError) {
+		switch distributorError := distributorError.(type) {
```
There's no need to do `errors.As(err, &distributorError)` if the only thing we doing next is a `switch` on the type and all of them are implementation of it.

duricanikolic replied:
If we just consider _this_ PR, it is true.
But in the next PR I am planning to add the `ingestererror.Error` interface, that will be used for the errors returned by the ingester. So I was thinking we could distinguish between them and handle them separately.

colega commented on pkg/util/push/push.go:
```diff
@@ -33,12 +34,27 @@ type bufHolder struct {
+	SkipLabelNameValidationHeader = "X-Mimir-SkipLabelNameValidation"
+	statusClientClosedRequest     = 499
+	// 529 is non-standard status code used by some services to signal that "The service is overloaded".
+	StatusServiceOverload = 529
```
```suggestion
	StatusServiceOverloaded = 529
```

colega commented on pkg/util/push/push.go:
```diff
@@ -121,15 +137,56 @@ func handler(maxRecvMsgSize int,
+	return http.StatusInternalServerError, err.Error()
+}
+
+func GetDistributorPushErrorHTTPStatus(distributorError distributor_error.DistributorError) int {
```
Shouldn't be exported, and should't have the `get` prefix.

```suggestion
func distributorPushErrorHTTPStatus(distributorError distributor_error.DistributorError) int {
```

colega commented on pkg/util/push/push.go:
```diff
@@ -121,15 +137,56 @@ func handler(maxRecvMsgSize int,
+
+func getHTTPStatusAndMessage(err error) (int, string) {
+	var distributorError distributor_error.DistributorError
+	if errors.As(err, &distributorError) {
+		msg := distributorError.Error()
+		code := GetDistributorPushErrorHTTPStatus(distributorError)
+		return code, msg
+	}
+	return http.StatusInternalServerError, err.Error()
```
Again: the second param is always `err.Error()`, and we don't need to check whether an error is `distributor_error.DistributorError` if we're checking specific implementations later.

duricanikolic replied:
In this case we will definitely receive both distributor and ingester errors, and I don't know yet whether we will need to distinguish them or not. But for the moment I can remove the `errors.As()`.

colega commented on pkg/distributor/distributor.go:
```diff
@@ -1093,11 +1093,43 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
+	var distributorError distributor_error.DistributorError
+	if errors.As(err, &distributorError) {
+		switch distributorError := distributorError.(type) {
+		case distributor_error.ReplicasNotMatchDistributorPushError:
```
Instead of checking specific types on errors, we should check `case errors.As(...):` as the error may come wrapped.

duricanikolic replied:
In the current implementation it is not the case, but it might be that in the future this will happen.
I will add this check.

colega commented on pkg/util/error/distributor.go:
```diff
@@ -0,0 +1,129 @@
+	return e.err
+}
+
+// PushError makes DistributorPushError implement pusherror.Error interface.
+func (e DistributorPushError) PushError() {
+}
+
+// DistributorError makes DistributorPushError implement DistributorError interface.
+func (e DistributorPushError) DistributorError() {
+}
```
Style nit, I think it could be opinonated for one liners above, but for empty methods I think that the newline removes readability:

```suggestion
// PushError makes DistributorPushError implement pusherror.Error interface.
func (e DistributorPushError) PushError() {}

// DistributorError makes DistributorPushError implement DistributorError interface.
func (e DistributorPushError) DistributorError() {}
```

pstibrany commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,127 @@
+	DistributorPushError
+}
+
+func NewReplicasNotMatchDistributorPushError(err error) ReplicasNotMatchDistributorPushError {
```
This is duplicate of `replicasNotMatchError`, we don't need another one.

duricanikolic replied:
Now that the package has been moved under `distributor/distributorerror` it makes sense. I can move `tooManyClustersError ` and `replicasNotMatchError` under `distributorerror`.

pstibrany commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,127 @@
+
+import "github.com/grafana/mimir/pkg/util/validation"
+
+type Error interface {
```
I'm not a fan of another `Error` interface with a method that does nothing. I see it's used to distinguish between "distributor" and "ingesters" errors, but still. Interface should define a behaviour.

colega replied:
After reviewing this PR 3 times, I agree here. There's no need for this interface. Distributor returns `error`, and some of them may have special meaning like `distributorerror.TooManyClusters`.

duricanikolic replied:
I have removed the `distributorerror.Error` interface.

duricanikolic commented on pkg/distributor/distributor.go:
```diff
@@ -728,18 +726,18 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
 			if errors.Is(err, tooManyClustersError{}) {
 				d.discardedSamplesTooManyHaClusters.WithLabelValues(userID, group).Add(float64(numSamples))
-				return httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+				return distributorerror.NewTooManyClustersDistributorPushError(err)
```
Now that the package has been moved under `distributor/distributorerror` it makes sense. I can move `tooManyClustersError ` and `replicasNotMatchError` under `distributorerror`.

colega commented on pkg/util/push/push.go:
```diff
@@ -121,15 +137,59 @@ func handler(maxRecvMsgSize int,
+			if ok {
+				code, msg = int(resp.Code), string(resp.Body)
+			} else {
+				code, msg = getHTTPStatusAndMessage(err)
```
```suggestion
				code, msg = distributorPushErrorHTTPStatus(err), err.Error()
```

And you can remove the `distributorPushErrorHTTPStatus` function.

colega commented on pkg/util/validation/validate.go:
```diff
@@ -373,7 +373,7 @@ type MetadataValidationConfig interface {
 
 // CleanAndValidateMetadata returns an err if a metric metadata is invalid.
-func CleanAndValidateMetadata(m *MetadataValidationMetrics, cfg MetadataValidationConfig, userID string, metadata *mimirpb.MetricMetadata) error {
+func CleanAndValidateMetadata(m *MetadataValidationMetrics, cfg MetadataValidationConfig, userID string, metadata *mimirpb.MetricMetadata) ValidationError {
```
Why this change?

duricanikolic replied:
@colega  `CleanAndValidateMetadata` returns only `ValidationError` errors, as expected, so I made it compliant with other validation methods from the package, that already return a `ValidationError`. This way, I could use the errors returned by all these methods inside the `NewValidationDistributorPushError()` constructor, that requires a `ValidationError` parameter.
If you don't agree with this change, I can change the signature of `NewValidationDsitributorPushError()`.

colega replied:
If you revert this change the code still compiles and works in the same way. This is because this interface doesn't add any extra information to the error: the interfaces in Go define behaviors and are implicitly implemented.

duricanikolic replied:
@colega I didn't get whether it is OK for you to do this change or you would prefer to revert it?

colega replied:
I would revert this change (and not related to this PR, but would also remove the `ValidationError` type at all) as it doesn't do anything. 

You don't really need to change the signature of `NewValidationDistributorPushError` because both `error` and `validation.ValidationError` describe the same thing: something with `Error() string` method.

duricanikolic replied:
OK, I have reverted it.

colega commented on pkg/util/push/push.go:
```diff
@@ -121,15 +137,59 @@ func handler(maxRecvMsgSize int,
-			if !ok {
-				http.Error(w, err.Error(), http.StatusInternalServerError)
-				return
+			if ok {
```
Can we make this, please, to make sure that `resp` only lives in the conditional block?
```suggestion
			if resp, ok := httpgrpc.HTTPResponseFromError(err); ok {
```

colega commented on pkg/util/push/push.go:
```diff
@@ -33,12 +34,27 @@ type bufHolder struct {
 
+// TODO: the idea of this interface is to represent the error interface of
+// push.Func, which would then have the following signature:
+// type Fund func(ctx context.Context, req *Request) Error.
```
I don't think that `push.Func` has to return a `push.Error`. There are a lot of errors that can happen which aren't something specific to `push.Func`.

I don't think we need this interface at all.

Also, nit:
```suggestion
// type Func func(ctx context.Context, req *Request) Error.
```

pstibrany replied:
> I don't think we need this interface at all.

Agree.

duricanikolic replied:
This interface has been removed.

duricanikolic commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,127 @@
+
+// ReplicasNotMatchDistributorPushError is an implementation of Error,
+// corresponding to distributor.replicasNotMatchError.
+type ReplicasNotMatchDistributorPushError struct {
```
I have replaced the `DistributorPushError` suffix with `Error` suffix.

colega replied:
As long as the package is called `distributorerror`, the `Error` suffix is also redundant, because the full name is:
`distributorerror.ReplicasNotMatchDistributorPush`.

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,130 @@
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// DistributorPushError is a generic error returned on
+// distributor's write path.
+type DistributorPushError struct {
+	err error
+}
```
Why do we still need this?

duricanikolic replied:
@colega 
Similarly as I explained above, my idea was to have all the errors returned by the distributor be of the same type.
This wrapper is now used for the following errors:
- `haTracker.updateKVStore()`
- `errMaxInflightRequestsReached`
- `errMaxIngestionRateReached`
- `errMaxInflightRequestsBytesReached`
They will all be treated as an internal server error. This wrapper would not add any additional benefits. 
But, since the last 3 are all instance-limit errors, maybe it wouldn't be bad to add another error type, `InstanceLimitError`, to group the last 3 errors together, and to remove `DistributorPushError`. WDYT?

colega replied:
> But, since the last 3 are all instance-limit errors, maybe it wouldn't be bad to add another error type, InstanceLimitError

Would they be treated in a special way or would they still be an internal server error?

duricanikolic replied:
Currently they would be translated int an internal server error.

duricanikolic replied:
I have removed the `NewDistributorPushError()` constructor and its usages, and I renamed `DistributorPushError` into `distributorPushError` and used it locally in `distributorerror/error.go` only.

colega replied:
Then I would not add an extra type, as it would not help.

colega commented on pkg/distributor/distributor.go:
```diff
@@ -1054,13 +1056,11 @@ func (d *Distributor) limitsMiddleware(next push.Func) push.Func {
+			return distributorerror.NewRequestRateError(
+				d.limits.RequestRate(userID),
+				d.limits.RequestBurstSize(userID),
+				d.limits.ServiceOverloadStatusCodeOnRateLimitEnabled(userID),
```
Can you please create an issue to move this decision to the API/Handler level? The error here should know nothing about the status code we want to return on one of the transport layers.

duricanikolic replied:
I have fixed this in the present PR.

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+	err error
+}
+
+// Error makes distributorPushError implement error interface.
+func (e distributorPushError) Error() string {
+	return e.err.Error()
+}
+
+// Unwrap makes distributorPushError implement Wrapper interface.
```
There's no need to add these comments on well-known interface implementations.

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+// used to represent all implementations of validation.ValidationError.
+type ValidationError struct {
+	distributorPushError
+	err string
```
If you keep this as is, this should be `msg string` (because `err` is used for `error` type and this is unexpected)

duricanikolic replied:
I have re-organized this implementation.

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// ValidationError is an implementation of Error,
+// used to represent all implementations of validation.ValidationError.
```
Note that again, all golang errors are implementations of validation.ValidationError :-/

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+	return e.err
+}
+
+// ReplicasNotMatchError is an implementation of Error,
```
This comment is outdated, there's no `Error` anymore.

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+
+// Error makes TooManyClustersError implement error interface.
+func (e TooManyClustersError) Error() string {
+	return globalerror.TooManyHAClusters.MessageWithPerTenantLimitConfig(
+		fmt.Sprintf("the write request has been rejected because the maximum number of high-availability (HA) clusters has been reached for this tenant (limit: %d)", e.limit),
+		validation.HATrackerMaxClustersFlag)
+}
```
```suggestion
	return globalerror.TooManyHAClusters.MessageWithPerTenantLimitConfig(
		fmt.Sprintf("the write request has been rejected because the maximum number of high-availability (HA) clusters has been reached for this tenant (limit: %d)", e.limit),
		validation.HATrackerMaxClustersFlag,
	)
}
```

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// IngestionRateError is an implementation of Error,
+// used to represent the ingestion rate limited error (globalerror.IngestionRateLimited).
```
Can we document somewhere what is the relation between `distributorerror` and `globalerror` packages? 

This package wraps several error values from that package with specific types here. Where do those live? I think the fact that those are both constants and error doesn't really help.

duricanikolic commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,121 @@
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type distributorPushError struct {
+	err error
+}
+
+// Error makes distributorPushError implement error interface.
+func (e distributorPushError) Error() string {
+	return e.err.Error()
+}
+
+// Unwrap makes distributorPushError implement Wrapper interface.
+func (e distributorPushError) Unwrap() error {
+	return e.err
+}
```
In the initial version of this PR, `Unwrap` was needed to keep the semantics of the underlying error, so for example, we don't lose the info that the error shouldn't be logged. This is, in fact, not needed anymore, since we abandoned the idea of wrapping all the errors by the `DistributorPushError`. I will remove this.

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// RequestRate is an error used to represent the request rate limited error.
+// In the error catalog, the ID of this error is globalerror.RequestRateLimited.
+type RequestRate struct {
```
Nit: just like `globalerror.RequestRateLimited` can this error be `RequestRateLimited`? Because "RequestRate" on it's own doesn't describe a failure, but a reason.

colega commented on pkg/util/validation/errors.go:
```diff
@@ -293,16 +293,20 @@ func NewMaxQueryExpressionSizeBytesError(actualSizeBytes, maxQuerySizeBytes int)
+func FormatRequestRateLimitedError(limit float64, burst int) string {
+	return globalerror.RequestRateLimited.MessageWithPerTenantLimitConfig(
 		fmt.Sprintf("the request has been rejected because the tenant exceeded the request rate limit, set to %v requests/s across all distributors with a maximum allowed burst of %d", limit, burst),
-		requestRateFlag, requestBurstSizeFlag))
+		requestRateFlag,
+		requestBurstSizeFlag,
+	)
 }
 
-func NewIngestionRateLimitedError(limit float64, burst int) LimitError {
-	return LimitError(globalerror.IngestionRateLimited.MessageWithPerTenantLimitConfig(
+func FormatIngestionRateLimitedError(limit float64, burst int) string {
```
Since you're touching this function names, I'd do: `FormatRequestRateLimitedErrorMessage` or simply `FormatRequestRateLimitedMessage`.
We're not returning an error here, but a message.

colega commented on pkg/util/validation/errors_test.go:
```diff
@@ -37,11 +37,11 @@ func TestNewTotalMaxQueryLengthError(t *testing.T) {
 
 func TestNewRequestRateLimitedError(t *testing.T) {
-	err := NewRequestRateLimitedError(10, 5)
-	assert.Equal(t, "the request has been rejected because the tenant exceeded the request rate limit, set to 10 requests/s across all distributors with a maximum allowed burst of 5 (err-mimir-tenant-max-request-rate). To adjust the related per-tenant limits, configure -distributor.request-rate-limit and -distributor.request-burst-size, or contact your service administrator.", err.Error())
+	err := FormatRequestRateLimitedError(10, 5)
+	assert.Equal(t, "the request has been rejected because the tenant exceeded the request rate limit, set to 10 requests/s across all distributors with a maximum allowed burst of 5 (err-mimir-tenant-max-request-rate). To adjust the related per-tenant limits, configure -distributor.request-rate-limit and -distributor.request-burst-size, or contact your service administrator.", err)
 }
 
 func TestNewIngestionRateLimitedError(t *testing.T) {
```
Can you update the test names too, please?

duricanikolic commented on pkg/api/api.go:
```diff
@@ -86,9 +87,10 @@ type API struct {
 }
 
-func New(cfg Config, serverCfg server.Config, s *server.Server, logger log.Logger) (*API, error) {
+func New(cfg Config, serverCfg server.Config, s *server.Server, limits *validation.Overrides, logger log.Logger) (*API, error) {
```
Note for reviewers: I've added `limits` as parameter here, because in the case of request rate limited error (new `RequestRateLimited`) the HTTP status code depends on the `-distributor.service-overload-status-code-on-rate-limit-enabled` CLI flag. The `limit` parameter is then passed to the constructor of the push handler.

pstibrany replied:
We don't need `limits` for API itself, only in `RegisterDistributor` method.

pstibrany replied:
And `RegisterIngester`.

duricanikolic replied:
I have fixed this. Thank you.

pracucci commented on pkg/distributor/distributor.go:
```diff
@@ -1093,11 +1089,56 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
+	case errors.As(err, &replicasNotMatchErr):
+		return httpgrpc.Errorf(http.StatusAccepted, err.Error())
+	case errors.As(err, &tooManyClusterErr):
+		return httpgrpc.Errorf(http.StatusTooManyRequests, err.Error())
```
Wasn't returning `StatusBadRequest` before? Now we're returning 429 which may be retried by the client, but this error shouldn't be retried because the retry will get the same error again.

pracucci commented on pkg/util/push/push.go:
```diff
@@ -121,15 +131,59 @@ func handler(maxRecvMsgSize int,
+	case errors.As(err, &replicasNotMatchErr):
+		return http.StatusAccepted
+	case errors.As(err, &tooManyClusterErr):
+		return http.StatusTooManyRequests
```
This needs to be fixed too.

duricanikolic replied:
I have fixed this

pracucci reviewed (approved):
I like the refactoring a lot. Improves very much how distributors errors are handled. I left few comments to address before merging tho.

pstibrany commented on pkg/util/push/push.go:
```diff
@@ -121,15 +131,59 @@ func handler(maxRecvMsgSize int,
 	})
 }
+
+func distributorPushErrorHTTPStatus(ctx context.Context, err error, limits *validation.Overrides) int {
```
I agree we should fix the duplicated logic.

Eventually (not in this PR) I think should modify `Distributor.Push` to not return `httpgrpc` errors at all.

pstibrany replied:
I think it's similar to ingester's suport for HTTP endpoint -- we only use it for testing, and we could drop it.

There is currently no component using that gRPC method. (No code uses `distributorpb.NewDistributorClient`)

pstibrany replied:
I believe in the past some of our enterprise proxies used it, but that's not the case anymore, and they send data via our internal gateway.

pstibrany replied:
Well, I was once again wrong. We use that client in our enterprise code.

duricanikolic replied:
>This function is duplicated with the same one in the distributor package. Why not moving it to distributorerror package? 

The idea was to remove any mention of HTTP from the distributor code, and to remove the remaining mapping present in the distributor in one of the following PRs.

duricanikolic replied:
>Well, I was once again wrong. We use that client in our enterprise code (for instrumentation). This usage could be replaced.

@pstibrany Do you mean to replace the usage of `distributorpb.NewDistributorClient()` from the enterprise code? To replace it with what?

pstibrany replied:
> @pstibrany Do you mean to replace the usage of `distributorpb.NewDistributorClient()` from the enterprise code? To replace it with what?

If we want to get rid of `distributor.Distributor` gRPC service (which I think would be nice, but outside of the scope of this PR), we would need to replace gRPC call to this service by one of HTTP call (prom push or otel).

duricanikolic replied:
I will take care of that in a followup PR

colega commented on pkg/distributor/distributorerror/error_test.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+func TestNewReplicasNotMatchError(t *testing.T) {
```
This errors isn't wrapping any error, it's a _root cause_. Do you mean wrapping this error in another one with `err := fmt.Errorf("something: w", distributorerror.ReplicasNotMatch)` and checking that `errors.As(err, &distributorerror.ReplicasNotMatch{})`? That is given by the implementation of `errors.As` and doesn't require extra methods on the error itself (the definition of `errors.As` is that the error is wrapping something assignable to what you pass on the second param)

pstibrany reviewed (commented):
I left some questions. We need to reduce logic duplication and fix error pointed out by Marco (I'd suggest adding test for it too, maybe into integrations tests)

duricanikolic commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,88 @@
+
+// ReplicasNotMatch is an error stating that replicas do not match.
+// This error is not exposed in the Error catalog.
+type ReplicasNotMatch struct {
```
By declaring `ReplicasNotMatch` as
```
type ReplicasNotMatch error
```
any error would be a `ReplicasNotMatch`.

duricanikolic replied:
By declaring `ReplicasNotMatch` (or any other error) as:
```
type ReplicasNotMatch error
```
any error would be an instance of that error. For example, 
```
err := NewErrorOfAnyType()
var replicasNotMatch ReplicasNotMatch
assert.ErrorAs(t, err, &replicasNotMatch)
```
would give no error. And this is not what we want.

colega replied:
Uh oh, race condition comment :)

pstibrany replied:
Thank you both for explanation. That makes sense. What still doesn't quite make sense to me is why do we need to wrap another error, and not just keep replicas (in `ReplicasNotMatch` case) and format error in `Error` message.

Or `TooManyClusters` with `limit`. Why all the extra error creation just to wrap it?

colega commented on pkg/distributor/distributorerror/error.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+func NewValidationError(err error) Validation {
+	return Validation{
```
Definitely, `errors.New(err.Error())` seems unnecessary, we should just do `return Validation{err}` as all we want is to give the semantic meaning of being a Validation error to that underlying error.

pstibrany replied:
Right, that would make sense to me in this case.

duricanikolic replied:
I have modified a bit this implementation. What I want to ensure here is that `validationErr.Error()` will call `validationErr.error.Error()`.

colega replied:
We still have `errors.New(err.Error())`, I still can't see the point.

colega replied:
The only reason why someone would want to do that, is to make an opaque new error, but we're embedding the error and not exposing the `Unwrap` method so it's opaque anyway.

duricanikolic commented:
> I left some questions. We need to reduce logic duplication and fix error pointed out by Marco (I'd suggest adding test for it too, maybe into integrations tests)

The test was actually already there, in `distributor_test`, but I have removed the status code comparison, since we wanted to remove HTTP status codes from the distributor. I have reverted the change, and now use the original version of `distributor_test.go`, with some small changes.

colega commented on pkg/distributor/distributorerror/errors.go:
```diff
@@ -0,0 +1,126 @@
+
+func NewReplicasNotMatchError(replica, elected string) ReplicasNotMatch {
+	return ReplicasNotMatch{
+		error: fmt.Errorf("replicas did not mach, rejecting sample: replica=%s, elected=%s", replica, elected),
```
```suggestion
		error: fmt.Errorf("replicas did not match, rejecting sample: replica=%s, elected=%s", replica, elected),
```

colega commented on pkg/util/validation/errors.go:
```diff
@@ -6,273 +6,158 @@
-		labelValue: labelValue,
-		series:     series,
-	}
+func newLabelValueTooLongError(series []mimirpb.LabelAdapter, labelValue string) distributorerror.Validation {
+	return distributorerror.NewValidationError(
+		errors.New(
+			globalerror.SeriesLabelValueTooLong.MessageWithPerTenantLimitConfig(
+				fmt.Sprintf("received a series whose label value length exceeds the limit, value: '%.200s' (truncated) series: '%.200s'", labelValue, formatLabelSet(series)),
+				maxLabelValueLengthFlag,
+			),
+		),
+	)
 }
```
Now it's more obvious than ever than all this code should live in the `distributor` package.

`distributor` calls `util/validation.ValidateLabels` that calls `distributor/distributorerror` :exploding_head: 

I would stop the scope creep here, and move that in a follow up PR.

colega commented on pkg/util/validation/errors.go:
```diff
@@ -293,16 +178,22 @@ func NewMaxQueryExpressionSizeBytesError(actualSizeBytes, maxQuerySizeBytes int)
-	return LimitError(globalerror.RequestRateLimited.MessageWithPerTenantLimitConfig(
-		fmt.Sprintf("the request has been rejected because the tenant exceeded the request rate limit, set to %v requests/s across all distributors with a maximum allowed burst of %d", limit, burst),
-		requestRateFlag, requestBurstSizeFlag))
+func NewRequestRateLimitedError(limit float64, burst int) distributorerror.RequestRateLimited {
+	format := globalerror.RequestRateLimited.MessageWithPerTenantLimitConfig(
+		"the request has been rejected because the tenant exceeded the request rate limit, set to %v requests/s across all distributors with a maximum allowed burst of %d",
+		requestRateFlag,
+		requestBurstSizeFlag,
+	)
+	return distributorerror.NewRequestRateLimitedError(format, limit, burst)
 }
```
Discussed offline: that depends on `requestRateFlag` which are defined here in this package. That would require moving more stuff. Let's keep it like that in this PR and then rethink.

duricanikolic replied:
I have found a workaround. @colega @pstibrany could you please have a look? Then if you prefer the previous proposal, I will revert the last commit.

colega commented on pkg/distributor/distributor.go:
```diff
@@ -728,15 +726,15 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
 		removeReplica, err := d.checkSample(ctx, userID, cluster, replica)
 		if err != nil {
-			if errors.Is(err, replicasNotMatchError{}) {
+			var replicasNotMatchErr distributorerror.ReplicasNotMatch
+			if errors.As(err, &replicasNotMatchErr) {
```
Since  you're still pushing changes, nit, this would be shorter and no need to pollute with another variable:
```suggestion
			if errors.As(err, &distributorerror.ReplicasNotMatch{}) {
```

colega commented on pkg/distributor/distributor.go:
```diff
@@ -728,15 +726,15 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
 			}
 
-			if errors.Is(err, tooManyClustersError{}) {
+			var tooManyClustersErr distributorerror.TooManyClusters
+			if errors.As(err, &tooManyClustersErr) {
```
Same:
```suggestion
			if errors.As(err, &distributorerror.TooManyClusters{}) {
```

colega commented on pkg/distributor/distributor.go:
```diff
@@ -1093,11 +1080,29 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
+	if err == nil {
+		serviceOverloadErrorEnabled = d.limits.ServiceOverloadStatusCodeOnRateLimitEnabled(userID)
+	}
+	httpStatus, ok := distributorerror.ToHTTPStatus(pushErr, serviceOverloadErrorEnabled)
+	if ok {
```
```suggestion
	if httpStatus, ok := distributorerror.ToHTTPStatus(pushErr, serviceOverloadErrorEnabled); ok {
```

This way `httpStatus` isn't accessible outside of the conditional scope.

colega commented on pkg/distributor/distributorerror/errors.go:
```diff
@@ -0,0 +1,150 @@
+	StatusServiceOverloaded = 529
+)
+
+// ReplicasNotMatch is an error stating that replicas do not match.
+type ReplicasNotMatch struct {
```
Nit, and we can do in another PR, but can we rename this error to `ReplicasDidNotMatch`. I didn't want to comment this in the first place, but since it's the third day reading the same error... :see_no_evil:

colega commented on pkg/distributor/distributorerror/errors.go:
```diff
@@ -0,0 +1,150 @@
+// ReplicasNotMatchErrorf formats and returns a ReplicasNotMatch error.
+// The error message is formatted according to the given format specifier, replica and elected.
+// In order to print the limit and the burst, the format specifier must contain two %s verbs.
+func ReplicasNotMatchErrorf(format, replica, elected string) ReplicasNotMatch {
+	return ReplicasNotMatch{
+		replica: replica,
+		elected: elected,
+		format:  format,
+	}
+}
```
I don't like the fact that this has to take a format that is compatible with the other params provided. Can we just take the `msg string` and assume that the entity that defined the `format` would format it?

colega replied:
Same applies to the rest of the errors in this package.

colega commented on pkg/distributor/distributorerror/errors.go:
```diff
@@ -0,0 +1,150 @@
+
+// TooManyClusters is an error stating that there are too many HA clusters.
+type TooManyClusters struct {
+	limit  int
+	format string
```
Nit, and probably will be refactored, but it _should_ be:
```suggestion
	format string
	limit  int
```
Because that's the order in which the params are defined in the constructor of the error and passes to fmt.Sprintf.

colega commented on pkg/util/validation/validate_test.go:
```diff
@@ -493,3 +531,14 @@ func TestMaxNativeHistorgramBuckets(t *testing.T) {
 }
+
+func tooManyLabelsArgs(series []mimirpb.LabelAdapter, limit int) []any {
+	metric := mimirpb.FromLabelAdaptersToMetric(series).String()
+	ellipsis := ""
+
+	if utf8.RuneCountInString(metric) > 200 {
+		ellipsis = "\u2026"
+	}
+
+	return []any{len(series), limit, metric, ellipsis}
```
Shouldn't we shorten the metric if it's too long?

duricanikolic replied:
This is the same implementation we are having on master right now ([here](https://github.com/grafana/mimir/blob/main/pkg/util/validation/errors.go#L103-L114))

colega replied:
Hum, okay, I'd say that's a bug.

pstibrany commented:
Problems I'd like to see fixed in subsequent PRs:

- dependency from `util/push` to `distributorerror`. Push handler is currently generic and it should not call error translation from distributor (sub)package directly. To solve this we could  pass "error translation function" to push handler. Later when ingester errors are supported too, this error translation function should handle both kinds of errors.

- `distributorerror` package -- all errors introduced here belong to `distributor` package. Furthermore they don't need to be exported. The only thing that needs to be exported is "distributor's error translation function".

- design of error types in `distributorerror` package. They should not be just wrappers around 'error' or 'format`.
